token/program-2022/src/extension/memo_transfer/instruction.rs

```diff
@@ -0,0 +1,116 @@
+use {
+    crate::{check_program_account, error::TokenError, instruction::TokenInstruction},
+    num_enum::{IntoPrimitive, TryFromPrimitive},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    std::convert::TryFrom,
+};
+
+/// Default Account State extension instructions
+#[derive(Clone, Copy, Debug, PartialEq, IntoPrimitive, TryFromPrimitive)]
+#[repr(u8)]
+pub enum MemoTransferInstruction {
+    /// Require memos for transfers into this Account. Adds the MemoTransfer extension to the
+    /// Account, if it doesn't already exist. In this case,
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The account to update.
+    ///   1. `[writeable, signer]` The payer account to fund realloc, if needed
+    ///   2. `[signer]` The account's owner.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The account to update.
+    ///   1. `[writeable, signer]` The payer account to fund realloc, if needed
+    ///   2. `[]` System program for realloc funding, if needed
```

Comment:
These are only needed when reallocation is required to make space for the extension... Currently, they are required for all Require instructions, even if the account won't need reallocation. It's not super clean, but we could make them optional. Thoughts?

Reply:
Does a 3rd instruction help? On `Init`, the account still may or may not be big enough to hold the new extension, so the payer and system program will still end up optional.

Reply:
What if we have a more general `Realloc` instruction, where the caller passes in a list of additional `ExtensionTypes` that it wants to ensure the account has space for.  The account owner doesn't even need to be a signer there, only the funding account.

Reply:
Oh, that seems nice. I think the account owner should still be a signer, though.
There might be reasons I don't want just anyone to be able to make my account huge, even if they pay the rent for it, eg. if tx-cost-model charges more load large accounts.

Reply:
I'll PR my vision of a `Realloc` instruction for discussion

Reply:
Awesome.

> There might be reasons I don't want just anyone to be able to make my account huge

I don't think I ever checked to see whether realloc allows account data to be shrunk?  But I guess I could still fund your reallocation through a system transfer, while you agree to it by actually calling the `Realloc` instruction

Reply:
> I don't think I ever checked to see whether realloc allows account data to be shrunk?

It does! https://github.com/solana-labs/solana/blob/28442aa9226907ceb926287906cc14135aafd18e/programs/bpf/rust/realloc/src/processor.rs#L78 has an example of that, so we have to be sure that reallocs are always larger, but I think the code already does that

Reply:
> I don't think I ever checked to see whether realloc allows account data to be shrunk?

I believe it can, yeah. I assume we would want to prevent shrinking in this `Realloc` instruction, though. (jinx, Jon)

Reply:
https://github.com/solana-labs/solana-program-library/pull/2864

Reply:
Reworked on ^